packages/platform-android/src/commands/runAndroid/index.ts

```diff
@@ -114,14 +114,30 @@ function getPackageNameWithSuffix(
 // Builds the app and runs it on a connected emulator / device.
 function buildAndRun(args: Flags, sourceDir: string) {
   const gradlew = process.platform === 'win32' ? 'gradlew.bat' : './gradlew';
+  const manifestPath = `${sourceDir}/${
+    args.appFolder
+  }/src/main/AndroidManifest.xml`;
+  let packageName;
+  let manifestContent;
 
-  // @ts-ignore
-  const packageName = fs
-    .readFileSync(
-      `${sourceDir}/${args.appFolder}/src/main/AndroidManifest.xml`,
-      'utf8',
-    )
-    .match(/package="(.+?)"/)[1];
+  try {
+    manifestContent = fs.readFileSync(manifestPath, 'utf8');
+  } catch (error) {
+    throw new CLIError(
+      "Failed to read AndroidManifest.xml. Maybe it's in a different Gradle project? If your project uses custom Android app directory, try `--appFolder` flag.",
+      error,
+    );
+  }
+
+  const packageMatch = manifestContent.match(/package="(.+?)"/);
+
+  if (packageMatch) {
+    packageName = packageMatch[1];
```

Comment:
One more thing, don't we already have a [`packageName`](https://github.com/react-native-community/cli/blob/next/packages/platform-android/src/config/index.ts#L43) in the configuration?

I would also consider returning `manifestPath` from the configuration as well, as [we already use it to derive `packageName`](https://github.com/react-native-community/cli/blob/next/packages/platform-android/src/config/index.ts#L41).

Otherwise, we risk using two different AndroidManifests which might be dangerous.

Reply:
Yes we have, but the tricky thing is the `appFolder`. Gradle has a concept of root project (where root build.gradle is) and sub projects (e.g. `app`, `core`, `ui`). In a typical RN app, in `settings.gradle` we only include `app` project. But technically, in an Android-first project, there could be many gradle sub projects – and hence `appFolder`. Our configuration doesn't have this as an option. We could add it, but even when overwritten through a flag, the `npx react-native config` won't know about it. This is likely a super edge-case or even not happening, but we still support it.

That's why I can't really use the packageName or even appId from the configuration – because they only refer to a single project.